test/Microsoft.AspNetCore.Mvc.Core.Test/Formatters/TextInputFormatterTest.cs

```diff
@@ -0,0 +1,167 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Text;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Http.Internal;
+using Microsoft.AspNetCore.Mvc.ModelBinding;
+using Xunit;
+
+namespace Microsoft.AspNetCore.Mvc.Formatters
+{
+    public class TextInputFormatterTest
+    {
+        [Fact]
+        public async Task ReadAsync_ReturnsFailure_IfItCanNotUnderstandTheContentTypeEncoding()
+        {
+            // Arrange
+            var formatter = new TestFormatter();
+            formatter.SupportedEncodings.Add(Encoding.ASCII);
+
+            var context = new InputFormatterContext(
+                new DefaultHttpContext(),
+                "something",
+                new ModelStateDictionary(),
+                new EmptyModelMetadataProvider().GetMetadataForType(typeof(object)),
+                (stream, encoding) => new StreamReader(stream, encoding));
+
+            context.HttpContext.Request.ContentType = "application/json;charset=utf-8";
+            context.HttpContext.Request.ContentLength = 1;
+
+            // Act
+            var result = await formatter.ReadAsync(context);
+
+            // Assert
+            Assert.Equal(true, result.HasError);
+            Assert.Equal(true, context.ModelState.ContainsKey("something"));
+            Assert.Equal(1, context.ModelState["something"].Errors.Count);
+
+            var error = context.ModelState["something"].Errors[0];
+            Assert.IsType<UnsupportedContentTypeException>(error.Exception);
```

Comment:
Discussed offline, it's not worth to add its own exception type given that this won't be a common scenario, our response will be 415 as with this exception type and the fact that we can easily change this in the future if we need to.